yarn-project/sequencer-client/src/block_builder/circuit_powered_block_builder.ts

```diff
@@ -381,6 +400,146 @@ export class CircuitPoweredBlockBuilder {
+        if (lowNullifier === undefined) {
+          return undefined;
+        }
+
+        const lowNullifierPreimage = new NullifierLeafPreimage(
+          new Fr(lowNullifier.value),
+          new Fr(lowNullifier.nextValue),
+          Number(lowNullifier.nextIndex),
+        );
+
+        // Get sibling path for existence of the old leaf
+        const siblingPath = await this.db.getSiblingPath(MerkleTreeId.NULLIFIER_TREE, BigInt(indexOfPrevious.index));
+
+        // Update the running paths
+        const witness = {
+          preimage: lowNullifierPreimage,
+          index: BigInt(indexOfPrevious.index),
+          siblingPath: siblingPath,
+        };
+        lowNullifierWitnesses.push(witness);
+
+        // Update subtree insertion leaf from null data
+        nullifierLeaf.nextIndex = lowNullifierPreimage.nextIndex;
+        nullifierLeaf.nextValue = lowNullifierPreimage.nextValue;
+
+        // Update the current low nullifier
+        lowNullifier.nextIndex = currInsertionIndex;
+        lowNullifier.nextValue = BigInt(newValue);
+
+        // Update the old leaf in the tree
+        await this.db.updateLeaf(MerkleTreeId.NULLIFIER_TREE, lowNullifier, BigInt(indexOfPrevious.index));
+      }
+
+      // increment insertion index
+      currInsertionIndex++;
+      insertionSubtree.push(nullifierLeaf);
+    }
+
+    // Create insertion subtree and forcefully insert in series
+    // Here we calculate the pointers for the inserted values, if they have not already been updated
+    for (let i = 0; i < leaves.length; i++) {
+      const newValue = new Fr(toBigIntBE(leaves[i]));
+
+      if (newValue.isZero()) continue;
+
+      // We have already fetched the new low nullifier for this leaf, so we can set its low nullifier
+      const lowNullifier = lowNullifierWitnesses[i].preimage;
+      // If the lowNullifier is 0, then we check the previous leaves for the low nullifier leaf
+      if (lowNullifier.leafValue.isZero() && lowNullifier.nextValue.isZero() && lowNullifier.nextIndex === 0) {
+        for (let j = 0; j < i; j++) {
+          if (
+            (insertionSubtree[j].nextValue > newValue && insertionSubtree[j].leafValue < newValue) ||
+            (insertionSubtree[j].nextValue.isZero() && insertionSubtree[j].nextIndex === 0)
+          ) {
+            insertionSubtree[j].nextIndex = Number(startInsertionIndex) + i;
+            insertionSubtree[j].nextValue = newValue;
+          }
+        }
+      }
+    }
+
+    // For each calculated new leaf, we insert it into the tree at the next position
+    for (let i = 0; i < insertionSubtree.length; i++) {
+      // We can skip inserting empty leaves
+
+      const asLeafData: LeafData = {
+        value: insertionSubtree[i].leafValue.value,
+        nextValue: insertionSubtree[i].nextValue.value,
+        nextIndex: BigInt(insertionSubtree[i].nextIndex),
+      };
+
+      await this.db.updateLeaf(MerkleTreeId.NULLIFIER_TREE, asLeafData, startInsertionIndex + BigInt(i));
```

Comment:
Can we use appendLeaf here instead? Not too important, I see updateLeaf updates the size anyway.

Reply:
unfortunately we cant, appending a leaf will fail as we have already forcefully updated pointers underneath in the previous step